Implement EvaluableExpression.Print

romankivalin opened the pull request:
Instead of using `evaluationStage`, I've added a new type `ExprNode`, which itself can be used in place of `EvaluableExpression`. It's actually an AST of the expression, but unlike `evaluationStage`, it doesn't contain built-in operators and exposes variable names and values. It's also much simpler: only tree node types: literal, variable, operator.

Based on the new `ExprNode` I've implemented print function, which can be used with a custom node handler, or use default node handler and override behavior with config (currently you can change how literals, variables and each operator are printed, map variable and operator names, override operator precedence).

Functions are not supported currently, because they are processed by parser, but we can change parser to create `ExprNode` directly, or write a new parser. What do you think, @smw1218, @palmdalian?

Still WIP, I want to add comments and write more tests.

palmdalian commented:
This one might be a little over my head, but maybe that's because I skipped the CS degree :)

I think the other Reduce PR is more in line with if we want to try to get changes back upstream to the main project, just because this one seems like a much bigger change. But if you think this is better and simpler for our use case, let's go for it!

Lack of functions shouldn't be too much of a problem right now since our expressions are fairly simple. Some tests/examples of the PrintConfig would be helpful so that I can drop it in to the timeline code. Nice work!

romankivalin commented:
@palmdalian I've reverted Eval and Reduce changes and added docs and test for Print. I can also plug this branch into timeline code, if you'd like.

I'm going to put Eval and updated Reduce on a separate branch and work on a new parser with functions support, but it looks like it wouldn't use any existing code, so it would be like a whole new library.
